StorageUtility/Linux/StorageUtility.cpp

```diff
@@ -17,251 +17,181 @@ limitations under the License.
-            return( eErrorCode::Memory );
-        }
-        pDevDetailData->cbSize = sizeof( SP_INTERFACE_DEVICE_DETAIL_DATA );
+    s_MappingSystemPathToProperties.clear();
+    udev_list_entry_foreach(dev_list_entry, devices)
+    {
+        const char *path;
+
+        /* Get the filename of the /sys entry for the device
+           and create a udev_device object (dev) representing it */
+        path    = udev_list_entry_get_name(dev_list_entry);
+        dev     = udev_device_new_from_syspath(udev, path);
+        String devNode = udev_device_get_devnode(dev);
+        sStorageAdapterProperty property;
 
-        if ( SetupDiGetDeviceInterfaceDetail( devInfoHandle, &devInterfaceData, pDevDetailData, bufferSize, NULL, NULL ) == FALSE )
+        if( true == InitStorageAdapterProperty(property, dev) )
         {
-            error = GetLastError();
-            if ( ERROR_INSUFFICIENT_BUFFER != error )
-            {
-                HeapFree(GetProcessHeap(), HEAP_NO_SERIALIZE, pDevDetailData);
-                if ( eOnErrorBehavior::Continue == OnErrorBehavior ) { continue; }
-                SetupDiDestroyDeviceInfoList( devInfoHandle );
-                return( eErrorCode::Unknown );
-            }
+            s_MappingSystemPathToProperties[devNode] = property;
+            ipPaths.push_back( devNode );
         }
 
-        eErrorCode errorCode;
-        errorCode = eErrorCode::Unknown;
-        Callback.Function(Callback.Data, devInfoHandle, devInfoData, devInterfaceData, pDevDetailData, bufferSize, errorCode);
-
-        HeapFree(GetProcessHeap(), HEAP_NO_SERIALIZE, pDevDetailData);
+        udev_device_unref(dev);
     }
+    /* Free the enumerator object */
+    udev_enumerate_unref(enumerate);
 
-    SetupDiDestroyDeviceInfoList( devInfoHandle );
-
-    return( eErrorCode::None );
-}*/
+    udev_unref(udev);
+    return( eErrorCode::None);
+}
 
-eErrorCode GetStorageAdapterProperty( HANDLE Handle, sStorageAdapterProperty& AdapterProperty )
+eErrorCode GetStorageDeviceHandle( const String& DevicePath, DeviceHandle& Handle )
 {
-   /* STORAGE_DESCRIPTOR_HEADER storageDescHeader;
-
-    STORAGE_PROPERTY_QUERY StorageProperty;
-    StorageProperty.PropertyId = StorageAdapterProperty;
-    StorageProperty.QueryType = PropertyStandardQuery;
-
-    DWORD bytesReturned;
-    DWORD ret =
-        DeviceIoControl(
-                        Handle,
-                        IOCTL_STORAGE_QUERY_PROPERTY,
-                        &StorageProperty,
-                        sizeof(STORAGE_PROPERTY_QUERY),
-                        &storageDescHeader,
-                        sizeof(STORAGE_DESCRIPTOR_HEADER),
-                        &bytesReturned,
-                        NULL
-                        );
-    if ( 0 == ret )
+    if ((Handle = open(DevicePath.c_str(), O_RDONLY|O_NONBLOCK)) < 0)
     {
-        return( eErrorCode::Io );
+        printf("ERROR: Cannot open device %s\n", DevicePath.c_str());
+        return eErrorCode::None;
```

Comment:
Please don't return error none here. ErrorNone means the operation is ok, so the out side will think "I get the DeviceHandle for this device path".